security/postmortems/cve-2019-9900.md

```diff
@@ -0,0 +1,321 @@
+library, adapted and minified for the Envoy context. The follow patches were
+produced:
+* https://github.com/envoyproxy/envoy/commit/c22cfd2c483fc26534382a0b6835f45264bb137a
+* https://github.com/envoyproxy/envoy/commit/7ed6d2187df94c4cb96f7dccb8643bf764af2ccb
+
+In both cases, the Envoy security team considered the issues of medium criticality (CVSS 6.5)
+initially, since it was thought that the attack complexity was high, requiring special circumstances
+to apply. As we continued discussion with Istio and Google teams, it became apparent that the
+exploits were trivial to automate and we upgraded to high criticality (CVSS 8.3), due to the lower
+attack complexity.
+
+A 1.9.1 security release was initially targeted for 2019-04-02 and announced on 2019-03-22. An
+e-mail was sent to the Envoy private distributor list sharing CVE details.
+
+After private discussions with a distributor on 2019-03-28, who expressed concern over the very
+short (3 working day) distance between fix patch availability and release, the Envoy security team
+decided to delay the 1.9.1 release until 2019-04-05. This provided 1 week for distributors to
+prepare their software for the security release date.
+
+Fix patches were shared with the private distributor list late on 2019-03-28.
+
+During the fix process, two distributors reached out to us to request the ability to stage in
+publicly accessible locations binary images with the fixes applied. While technically this would
+violate embargo, we decided to allow this due to a lack of a clear alternative; Envoy's sidecar
+use cases and reliance on Docker for distribution, where images are generally staged on public hubs,
+did not lend itself to opaque rollout.
+
+## Detection
+
+The underlying issue behind CVE-2019-9900 was first noticed via fuzzers when an explicit `ASSERT`
+check for embedded NULs was added in #6170. The following issue was tripped by
+`h1_capture_fuzz_test`:
+https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=13613. Some experiments with `netcat`, `tcpdump`
+and an Envoy binary demonstrated that it was viable to at least bypass header suffix matches via
+this mechanism.
+
+CVE-2019-9901 was reported by an external researcher (Erlend Oftedal) to the Istio security team.
+
+## Action Items
+
+* https://github.com/envoyproxy/envoy/issues?utf8=%E2%9C%93&q=is%3Aissue+%22Action+item+for+CVE-2019-9900%22+
+* https://github.com/envoyproxy/envoy/issues?utf8=%E2%9C%93&q=is%3Aissue+%22Action+item+for+CVE-2019-9901%22+
+
+## Lessons Learned
+
+### What went well
+
+* Fix patches were available within 2 weeks of vulnerability disclosure. The
+  changes were localized and relatively clean.
+
+### What went wrong
+
+* The Envoy private distributor list was initially almost empty. We sent out an
+  e-mail to remind distributors to sign up on 2019-03-14 and the list is now O(10).
+
+* The security impact of https://github.com/envoyproxy/envoy/issues/6008
```

Comment:
Maybe add a comment about the fact that https://github.com/envoyproxy/envoy/issues/2956 was opened a year earlier and also wasn't caught as a potential security issue? This is mentioned in the timeline, but that's a section that's usually skipped.